### PR TITLE
Keep post publishing popover open when a date is clicked

### LINKED ISSUE #24206: DateTimePicker: Click on date causes popover to close automatically
**Describe the bug**
I'm trying to use the `DateTimePicker` component inside of a `Dropdown` component with other settings. Unfortunately, clicking on a date in the picker automatically closes the popover. This prevents the user to change other settings unless they reopen the popover.

The same happens with the default date picker to schedule a post. It might be the case that this is intentional but it only "works" as is due to the way how the `DayPickerSingleDateController` and `Popover` components are working.

**To reproduce**
1. Create a new post
2. Open the schedule popover by clicking on the publish date 
3. Change the date or month by using the input/dropdown menu
4. Notice that the popover does not close
5. Now click on a date in the picker
6. The popover automatically closes

See also https://wordpress.slack.com/archives/C02QB2JS7/p1592578807492400 for a gif.

**Expected behavior**
The popover stays open so I can review the selected data in case I clicked on a wrong date. Or, when used in a custom component, `DateTimePicker` should have a setting to prevent this behaviour.

**Additional context**
This happens because when clicked on a date in `DayPickerSingleDateController`, `document.activeElement` it set to the `body` element.
[`PostSchedule`](https://github.com/WordPress/gutenberg/blob/b9de73ddb58b6b88faa4846d3f8294c815c9d4df/packages/editor/src/components/post-schedule/index.js#L27-L32) with `DateTimePicker` is used in a [`Dropdown`](https://github.com/WordPress/gutenberg/blob/b9de73ddb58b6b88faa4846d3f8294c815c9d4df/packages/edit-post/src/components/sidebar/post-schedule/index.js#L17-L33) component which sets the [`onFocusOutside` prop](https://github.com/WordPress/gutenberg/blob/b9de73ddb58b6b88faa4846d3f8294c815c9d4df/packages/components/src/dropdown/index.js#L93) with a function that checks [if the `document.activeElement` is part of the container ref](https://github.com/WordPress/gutenberg/blob/b9de73ddb58b6b88faa4846d3f8294c815c9d4df/packages/components/src/dropdown/index.js#L58-L72). Since the `body` element fails that check so the popover closes.

The `document.activeElement` issue can also be reproduced on react-dates' storybook: https://airbnb.io/react-dates/iframe.html?id=daypickersingledatecontroller--default

### PR DESCRIPTION
**Pull request by allilevine**
## Description

Add a keepOpen prop to the DateTimePicker and keep focus on the date picker popover when true.

This isn't an ideal solution, given [the note to remove](https://github.com/WordPress/gutenberg/blob/86a9dc74cce28da4b5f43123e4d57a2982d203fb/packages/components/src/date-time/date.js#L31) `keepFocusInside()`, but removing it would require either changes to [react-dates](https://github.com/airbnb/react-dates) or [replacing react-dates](https://github.com/WordPress/gutenberg/issues/21820). 
- Open `react-dates` issue to keep focus after user changes month: https://github.com/airbnb/react-dates/issues/1065
- There's a prop [`isFocused`](https://github.com/airbnb/react-dates/blob/6c2cb61c5e876fbdac72e6efed49c0bd64f06d6b/src/components/DayPickerSingleDateController.jsx#L96) we can pass to `<DayPickerSingleDateController>`, but the focus doesn't persist after transitioning to a different month. 

To do: Open a follow-up issue to fix the popover changing position as the date changes (the popover position is relative to the publish date width).

Fixes #24206 

## How has this been tested?

    1. Create a new post
    2. Open the schedule popover by clicking on the publish date
    3. Click on a date in the picker and confirm that the popover does not close.
    4. Navigate to the next month
    5. Click on a date and confirm that the popover does not close.
    6. Navigate to a date via the keyboard and select it. Confirm that the popover does not close.

## Screenshots 
Gif of the popover staying open after clicking on a date:
datetimepickerstaysopen

## Types of changes

Bug fix: adds a `keepOpen` prop to `DateTimePicker` and uses `keepFocusInside()` to keep the popover open when true.

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

File: packages/components/src/date-time/index.js [hunk 1]
```diff
@@ -28,6 +28,7 @@ function DateTimePicker(
 		onMonthPreviewed,
 		onChange,
 		events,
+		keepOpen,
```
Comment (retrofox) [inline packages/components/src/date-time/index.js:31]: Should `keepOpen` be defined `true` as default? where else the `<DateTimePicker />` is rendered?

Comment (allilevine) [reply packages/components/src/date-time/index.js:31]: I thought it was being used by other features, but it's just Storybook components. 😂 So it probably does make sense to default to true.

Comment (retrofox) [reply packages/components/src/date-time/index.js:31]: I think it's ok keeping it as `false` as default, actually. So by default, it won't change anything. Is it correct?

Comment (allilevine) [reply packages/components/src/date-time/index.js:31]: Yes, that's correct. If the `keepOpen` prop isn't passed to `DateTimePicker` or isn't `true` then there won't be any change to the popover behavior.

**Review by retrofox — APPROVED** [review]
LGTM. Although it might be fixed straight in the react-dates library, [it seems that we aren't going to get a fix](https://github.com/airbnb/react-dates/issues/1065). We can re-iterate and update the code later in case the issue is fixed there.
Also, we've been considering replacing the component with other choices.
I'd like to know the opinion of other folks, too.
This PR fixes the issue, which is nice.

**Review by tomjn — APPROVED** [review]
This looks sensible, I'm giving this a thumbs up 👍🏻

**Comment by talldan** [comment]
Looks good, thanks for working on this, I renamed the PR to make it a bit more user-oriented, as the title ends up in the changelog for the release.

File: packages/components/src/date-time/README.md [hunk 2]
```diff
@@ -84,3 +84,10 @@ This function will be called on each day, every time user browses into a differe
 
 - Type: `Function`
 - Required: No
+
+### keepOpen
```
Comment (youknowriad) [inline packages/components/src/date-time/README.md:88]: Thanks for the fix, I have one question here.

Do we really want this prop? I mean it seems sensible to always keep the focus in the date time picker regardless of whether it's inside a popover or not.

Comment (retrofox) [reply packages/components/src/date-time/README.md:88]: yeah, that's something we talked about a little bit [here](https://github.com/WordPress/gutenberg/pull/29738#pullrequestreview-609745144). Maybe we should remove the prop.

Comment (youknowriad) [reply packages/components/src/date-time/README.md:88]: Anyone for a follow-up here ;) ?

Comment (retrofox) [reply packages/components/src/date-time/README.md:88]: I think @allilevine will be happy to take it. Otherwise, I can take over too.

Comment (allilevine) [reply packages/components/src/date-time/README.md:88]: Yep I can work on a follow-up to remove the prop.

Comment (allilevine) [reply packages/components/src/date-time/README.md:88]: Follow-up here: https://github.com/WordPress/gutenberg/pull/29893

**Comment by allilevine** [comment]
I opened a follow-up issue to fix the popover position changing as the date changes: https://github.com/WordPress/gutenberg/issues/29899

**Pull request merged**